src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -0,0 +1,938 @@
+
+        private static string Format(DateTime value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(DateTime? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(decimal value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(decimal? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(double value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(double? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(float value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(float? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(Guid value)
+            => value.ToString(null, CultureInfo.InvariantCulture);
+
+        private static string? Format(Guid? value)
+            => value?.ToString(null, CultureInfo.InvariantCulture);
+
+        private static string Format(int value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(int? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        private static string Format(long value)
+            => value.ToString(CultureInfo.InvariantCulture);
+
+        private static string? Format(long? value)
+            => value?.ToString(CultureInfo.InvariantCulture);
+
+        // Used for constructing a URI with a new querystring from an existing URI.
+        private struct QueryStringBuilder
+        {
+            private readonly StringBuilder _builder;
+
+            private bool _hasNewParameters;
+
+            public string UriWithQueryString => _builder.ToString();
+
+            public QueryStringBuilder(ReadOnlySpan<char> uriWithoutQueryString, int additionalCapacity = 0)
+            {
+                _builder = new(uriWithoutQueryString.Length + additionalCapacity);
+                _builder.Append(uriWithoutQueryString);
+
+                _hasNewParameters = false;
+            }
+
+            public void AppendParameter(ReadOnlySpan<char> encodedName, ReadOnlySpan<char> encodedValue)
+            {
+                if (!_hasNewParameters)
+                {
+                    _hasNewParameters = true;
+                    _builder.Append('?');
+                }
+                else
+                {
+                    _builder.Append('&');
+                }
+
+                _builder.Append(encodedName);
+                _builder.Append('=');
+                _builder.Append(encodedValue);
+            }
+        }
+
+        // A utility for feeding a collection of parameter values into a QueryStringBuilder.
```

Comment:
Might be great to expand the comment to say something like "This is used when <case>"